src/MainWindow.vala

```diff
@@ -311,4 +314,44 @@ public class AppCenter.MainWindow : Gtk.ApplicationWindow {
         category_view.return_clicked ();
         trigger_search ();
     }
+
+    private void on_operation_finished (AppCenterCore.Package package, AppCenterCore.Package.State operation, Error? error) {
+        switch (operation) {
+            case AppCenterCore.Package.State.INSTALLING:
+                if (error == null) {
+                    // Check if window is focused
+                    var win = get_window ();
+                    if (win != null && (win.get_state () & Gdk.WindowState.FOCUSED) != 0) {
+                        break;
+                    }
+
+                    var notification = new Notification (_("Application installed"));
+                    notification.set_body (_("%s has been successfully installed").printf (package.get_name ()));
+                    notification.set_icon (new ThemedIcon ("system-software-install"));
+                    notification.set_default_action ("app.open-application");
+
+                    var app = get_application ();
+                    if (app != null) {
+                        app.send_notification ("installed", notification);
+                    }                    
+                } else {
+                    // Check if permission was denied or the operation was cancelled
+                    if (error.matches (IOError.quark (), 19) || error.matches (Pk.ClientError.quark (), 303)) {
+                        break;
+                    }
+
+                    var close_button = new Gtk.Button.with_label (_("Close"));
+
+                    var dialog = new MessageDialog (_("Application installation failed"), _("There was an error installing %s:\n%s").printf (package.get_name (), error.message), "dialog-error");
```

Comment:
Please use Title Case here for the primary text.

Is there a reason that we're not using the more specific "There was an error installing %s" for the primary text? Seems like that would flow better and then just use the error message as secondary text.